### PR TITLE
Fix Pet filter bug in staff view

### PR DESCRIPTION
**Pull request by khiga8**
# 🔗 Issue

https://github.com/rubyforgood/pet-rescue/issues/969

# ✍️ Description

This PR addresses a bug in the search filter resulting in the `unadopted` filter to not work as expected.

Changes made:
* Defines a missing `unadopted` scope.
* Update the `ransack_adopted` method to use strings instead of booleans to get around a known bug. Kudos to @coalest for pointing me to https://github.com/activerecord-hackery/ransack/issues/1375#issuecomment-1674885632.
* Add test coverage for `ransack_adopted` to ensure the adopted/unadopted filters work.
* Add missing test coverage for other filters in the search.

### CODE REVIEW

File: test/models/pet_test.rb [hunk 1]
```diff
@@ -126,4 +126,38 @@ class PetTest < ActiveSupport::TestCase
...
+    should "should return adopted pets via Pet.ransack" do
+      adopted_pets = Pet.ransack({"ransack_adopted"=>"true"})
+
+      assert_equal Pet.ransack_adopted(true).to_sql, adopted_pets.result.to_sql
+      assert_equal 1, adopted_pets.result.count
+    end
+
+    # Fails. SQL isn't what we'd expect?
+    should "should return unadopted pets via Pet.ransack" do
+      unadopted_pets = Pet.ransack({"ransack_adopted"=>"false"})
```
Comment (khiga8) [inline test/models/pet_test.rb:157]: This isn't returning the SQL we'd expect. In other words, it's not calling the else condition of `def self.ransack_adopted(boolean)`. 

Perhaps class methods with conditional logic isn't supported by `ransackable_scopes` 🤔

Comment (kasugaijin) [reply test/models/pet_test.rb:157]: @khiga8 sorry, I did not get notified when you made this comment...only just saw it. 

It's currently a bit weird. The method returns a scope depending on the argument, which is OK. But, we have an `adopted` scope defined on `Pet`, but I do not see one defiend for `unadopted`. Can you try creating an `unadopted` scope and see if this filter behaves as expected?  For the unadopted scope, we will want to find Pets that do not have a Match with `type` `adoption`. All other pets should be included in the results (including those with a Match `type` of `foster`.
 
```
  def self.ransack_adopted(boolean)
    boolean ? adopted : unadopted # there is no unadopted scope on the Pet model
  end
```

**Comment by kasugaijin** [comment]
@all-contributors
please add @khiga8 for code.

File: app/models/pet.rb [hunk 2]
```diff
@@ -132,7 +132,7 @@ def self.ransackable_scopes(auth_object = nil)
   end
 
   def self.ransack_adopted(boolean)
-    boolean ? adopted : unadopted
+    boolean ? adopted : Pet.where.not(id: Pet.adopted)
```
Comment (kasugaijin) [inline app/models/pet.rb:135]: Sorry, I see you have already done this! Can we make this a scope called `unadopted` instead?

File: test/models/pet_test.rb (hunk 1 above)
Comment (kasugaijin) [reply test/models/pet_test.rb:157]: Ok I get what you're saying now I am caught up. So, we have asserted the expected behaviour of this filter in this file, right? `test/controllers/organizations/staff/pets_controller_test.rb` 

And you have tested this works as expected in the browser?

If so, I am comfortable removing these tests. I don't think we need to assert the SQL. Thoughts?

File: app/models/pet.rb (hunk 2 above)
Comment (khiga8) [reply app/models/pet.rb:135]: I can do that once I get the tests passing! Still having problems with this. 😅

File: test/models/pet_test.rb (hunk 1 above)
Comment (kasugaijin) [reply test/models/pet_test.rb:157]: I can check out your branch this week to see but before I can get to that can you check the pet object you create in the setup to make sure it belongs to the expected organization

**Comment by kasugaijin** [comment]
@khiga8 how's this going? Do you have any questions?

File: test/models/pet_test.rb (hunk 1 above)
Comment (khiga8) [reply test/models/pet_test.rb:155]: Hey @kasugaijin. The Pet object within my test all belongs to the same organization.

I've updated the tests so it's clearer what cases are failing. 

The following scenarios work as expected (as shown by the passing tests):
* `Pet.ransack_adopted(true)`
* `Pet.ransack_adopted(false)`
* `Pet.ransack({ "ransack_adopted" => "true" })

However, the following scenario fails:
* `Pet.ransack({ "ransack_adopted" => "false" })

Curious if you might know why  `Pet.ransack_adopted(false)` returns expected results, but `Pet.ransack({ "ransack_adopted" => "false" })` doesn't.

Comment (coalest) [reply test/models/pet_test.rb:155]: @khiga8 I think you are running into an open issue/bug with the ransack gem. This workaround might help: https://github.com/activerecord-hackery/ransack/issues/1375#issuecomment-1674885632

**Comment by kasugaijin** [comment]
@all-contributors
please add @coalest for review.

File: test/models/pet_test.rb (hunk 1 above)
Comment (khiga8) [reply test/models/pet_test.rb:155]: Amazing @coalest, thank you for pointing me to that 😁

File: app/models/pet.rb (hunk 2 above)
Comment (khiga8) [reply app/models/pet.rb]: Done :)

**Comment by khiga8** [comment]
Hi! This is now ready for a review, thank you!

File: app/models/pet.rb [hunk 3]
```diff
@@ -131,8 +132,9 @@ def self.ransackable_scopes(auth_object = nil)
     [:ransack_adopted, :ransack_birth_date]
   end
 
-  def self.ransack_adopted(boolean)
-    boolean ? adopted : unadopted
+  # Using string values to get around ransack bug: https://github.com/activerecord-hackery/ransack/issues/1375
```
Comment (kasugaijin) [inline app/models/pet.rb:135]: Nice, thank you for leaving a comment on this.

**Review by kasugaijin — APPROVED** [review]
LGTM. Thankyou for getting to the bottom of this one!

**Comment by kasugaijin** [comment]
@khiga8 could just please run `rails standard:fix` to lint your changes - then it should pass the pipeline.

**Review by kasugaijin — APPROVED** [review]

**Pull request merged**